File: Java/src/Abilities.java
package net.glitchtechs.starapi;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.entity.*;
import org.bukkit.inventory.ItemStack;
import org.bukkit.util.Vector;

import java.util.List;

public class Abilities {

    public static void teleportPlayer(Player player, double distance) {
        Location playerLocation = player.getLocation();
        World world = player.getWorld();

        Vector direction = playerLocation.getDirection().multiply(distance);
        Location destination = playerLocation.add(direction);

        Location targetLocation = getSafeDestination(destination);
        if (targetLocation != null) {
            player.teleport(targetLocation);
        }
    }

    public static void createExplosion(Location center, double radius, double damage) {
        World world = center.getWorld();

        List<Entity> nearbyEntities = (List<Entity>) world.getNearbyEntities(center, radius, radius, radius);
        for (Entity entity : nearbyEntities) {
            if (entity instanceof LivingEntity && !(entity instanceof Player)) {
                ((LivingEntity) entity).damage(damage);
            }
        }

        world.createExplosion(center.getX(), center.getY(), center.getZ(), 0.0f, false, false);
    }

    public static void multiShot(Player player, int arrowAmount, double damage, boolean isCritical){
        arrowAmount = Math.max(0, arrowAmount);

        ItemStack arrowStack = new ItemStack(Material.ARROW, arrowAmount);

        if (!player.getInventory().containsAtLeast(arrowStack, arrowAmount)) {
            return;
        }
        player.getInventory().removeItem(arrowStack);

        for (int i = 0; i < arrowAmount; i++){
            Arrow arrow = player.launchProjectile(Arrow.class, player.getLocation().getDirection());
            arrow.setPickupStatus(AbstractArrow.PickupStatus.DISALLOWED);

            arrow.setDamage(damage);
            arrow.setCritical(isCritical);
        }
    }


    private static Location getSafeDestination(Location destination) {
        World world = destination.getWorld();
        Vector direction = destination.getDirection();

        for (double i = 0.0; i < 1.0; i += 0.1) {
            Location checkLocation = destination.clone().add(direction.clone().multiply(i));
            if (checkLocation.getBlock().getType() != Material.AIR) {
                return checkLocation.subtract(direction.clone().multiply(0.2));
            }
        }
        return destination;
    }

}

File: Kotlin/src/Abilities.kt
package net.glitchtechs.starapi

import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.*
import org.bukkit.inventory.ItemStack
import kotlin.math.max


object Abilities {
    fun teleportPlayer(player: Player, distance: Double) {
        val playerLocation = player.location
        val world = player.world

        val direction = playerLocation.direction.multiply(distance)
        val destination = playerLocation.add(direction)

        val targetLocation = getSafeDestination(destination)
        if (targetLocation != null) {
            player.teleport(targetLocation)
        }
    }

    fun createExplosion(center: Location, radius: Double, damage: Double) {
        val world = center.world

        val nearbyEntities = world!!.getNearbyEntities(center, radius, radius, radius) as List<Entity>
        for (entity in nearbyEntities) {
            if (entity is LivingEntity && entity !is Player) {
                entity.damage(damage)
            }
        }

        world.createExplosion(center.x, center.y, center.z, 0.0f, false, false)
    }

    fun multiShot(player: Player, arrowAmount: Int, damage: Double, isCritical: Boolean) {
        var arrowAmount = arrowAmount
        arrowAmount = max(0.0, arrowAmount.toDouble()).toInt()

        val arrowStack = ItemStack(Material.ARROW, arrowAmount)

        if (!player.inventory.containsAtLeast(arrowStack, arrowAmount)) {
            return
        }
        player.inventory.removeItem(arrowStack)

        for (i in 0 until arrowAmount) {
            val arrow = player.launchProjectile(Arrow::class.java, player.location.direction)
            arrow.pickupStatus = AbstractArrow.PickupStatus.DISALLOWED

            arrow.damage = damage
            arrow.isCritical = isCritical
        }
    }


    private fun getSafeDestination(destination: Location): Location {
        val world = destination.world
        val direction = destination.direction

        var i = 0.0
        while (i < 1.0) {
            val checkLocation = destination.clone().add(direction.clone().multiply(i))
            if (checkLocation.block.type != Material.AIR) {
                return checkLocation.subtract(direction.clone().multiply(0.2))
            }
            i += 0.1
        }
        return destination
    }
}